### PR TITLE
Add serailze methods to documents

### PR DESCRIPTION
**Pull request by JordonPhillips**
This adds a serialize method to documents and fixes a few small issues that cropped up when testing those changes. Notably an is_null method was added since it became apparent that otherwise there's no way to tell. More importantly though, document members no longer resolve to their target shapes, but rather the proper member shapes. This necessitated some changes to setitem. Lastly, some improvements were added to shape type inference that leave the result as document only when the value is null.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by JordonPhillips** [comment]
Hold up on reviewing this, I need to make changes

File: python-packages/smithy-core/smithy_core/documents.py [hunk 1]
```diff
@@ -86,20 +86,19 @@ def __init__(
                 self._raw_value = value
             else:
                 self._value = value  # type: ignore
-        elif isinstance(value, Sequence):
+        elif isinstance(value, Sequence) and not isinstance(value, str | bytes):
```
Comment (nateprewitt) [inline python-packages/smithy-core/smithy_core/documents.py:89]: This is the first time I've seen this. Was this a change with typing that enabled usage of this syntax vs a tuple of types?

File: python-packages/smithy-core/smithy_core/documents.py [hunk 2]
```diff
@@ -212,17 +217,15 @@ def as_map(self) -> dict[str, "Document"]:
         )
 
     def _wrap_map(self, value: Mapping[str, DocumentValue]) -> dict[str, "Document"]:
-        if self._schema.shape_type is not ShapeType.STRUCTURE:
+        if self._schema.shape_type not in [ShapeType.STRUCTURE, ShapeType.UNION]:
```
Comment (nateprewitt) [inline python-packages/smithy-core/smithy_core/documents.py:220]: nit
```suggestion
        if self._schema.shape_type not in (ShapeType.STRUCTURE, ShapeType.UNION):
```

**Review by nateprewitt — COMMENTED** [review]
Question and a minor nit, otherwise looks good to me.

File: python-packages/smithy-core/smithy_core/documents.py (hunk 1 above)
Comment (JordonPhillips) [reply python-packages/smithy-core/smithy_core/documents.py:89]: Yeah, [PEP604](https://peps.python.org/pep-0604/), which added the union syntax, also made it useable in isinstance checks.

**Review by nateprewitt — APPROVED** [review]

**Pull request merged**